### DISCUSSION TITLE
How long does your typical pull request pipeline take to run

### DISCUSSION THREAD

**Discussion by prolificcoder** [MobileNativeFoundation/discussions#199] [Polls]
Wanted to get a pulse on CI pipeline efficiencies for mobile code bases. 

I am assuming you are doing at least lint, unit tests, ui tests as part of your pipeline but not deployment to app stores or internal releases. It can be either Android, iOS, RN or Flutter, would be good to know what platform though. 

Our pipelines are bordering 45 mins and compared to backend that seems very slow.

**Comment by keith** [comment]
Ours is in the fastest bucket here, but at least some difference compared to your explanation is that we don't run blocking UI tests. Developers can opt into running them if they'd like, but otherwise they're just run on a cron and we debug the breaking change async

**Comment by saad352** [comment]
We have 2 apps. One app have an avg time of `~7 mins` with p90 of `9.35 mins`. Other one has an avg of `~11 mins` with p90 of `13.7 mins`. We also don't run UITest on PR checks right now. But we soon plan to run them without effecting the CI time. (Running them in parallel)

The main reason for the big difference between avg and p90 is that we run selected tests depending upon the changes user has introduced. So tests are run for those modules which are effected by the changes in PR.

**Comment by 4itdg3x9** [comment]
Lack of Remote Caching: If you aren't using Gradle Remote Build Cache (Android) or Bazel, you are wasting time recompiling unchanged modules.
Serial UI Testing: Never run these sequentially. You need a device farm (Firebase Test Lab, BrowserStack) to shard your UI tests across 10+ devices simultaneously. If they run one after another, your pipeline will never be fast.
The "Clean Build" Habit: Ensure your CI runners are persisting the build cache across runs. If you start [Vegasnow Casino](https://vegasnowcasino-au.net/) every pipeline with a ./gradlew clean, you’ve already lost 10 minutes.
Test Impact Analysis: You shouldn't be running 100% of your tests for every minor PR. Use tools to detect affected modules and only run tests relevant to the changed code.
